src/main.cpp

```diff
@@ -2363,11 +2363,128 @@ bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockIndex* pin
+        // occured in a prior block.
+        const CCoins* prev_final_tx = view.AccessCoins(view.GetFinalTx());
+        if (!prev_final_tx) {
+            // This state of affairs can potentially happen when performing an
+            // upgrade from a prior node version to one which supports
+            // validation of block-final transaction rules, but *after* rule
+            // activation. Normal tip validation in init.cpp will halt on this
+            // error and notify that the user that their block database is
+            // corrupted, which is fixed by starting with -reindex=1.
+            return state.DoS(100, error("%s: prior block-final tx hash not found; corruption likely!", __func__),
+                             REJECT_INVALID, "corrupt-db-no-blockfinal-hash", true);
+        }
+        // The output spent by the very first block-final transaction is
+        // generated in a coinbase at the point of activation, so it takes
+        // COINBASE_MATURITY blocks until that output matures and we're able to
+        // really start enforcing the block-final tx rules.
+        if (!prev_final_tx->IsCoinBase() || (pindex->nHeight - prev_final_tx->nHeight >= COINBASE_MATURITY)) {
+            // Make sure the block contains *at least* 2 transactions, the
+            // coinbase and the block-final transaction. This is not stricly
+            // necessary since coinbase can't have inputs, but for clarity we
+            // explicitly check and fail with the relevant error message.
+            if (block.vtx.size() < 2) {
+                return state.DoS(100, error("%s: missing block-final transaction", __func__),
+                                 REJECT_INVALID, "missing-block-final-tx");
+            }
+            const CTransaction& final_tx = block.vtx.back();
+            // Make sure each txin comes from either the prior block-final
+            // transaction, or an output that matured in this block.
+            std::size_t spends_prior_tx = 0;
+            BOOST_FOREACH(const CTxIn& txin, final_tx.vin) {
+                // All outputs of the prior block-final transaction need to be
+                // spent, so obviously they're allowed.
+                if (txin.prevout.hash == view.GetFinalTx()) {
+                    ++spends_prior_tx;
+                    continue;
+                }
+                // Fetching the UTXO record will fail for any outputs sourced
+                // from the current block, and all such spends are allowed (for
+                // future expansion purposes):
+                const CCoins* from = view.AccessCoins(txin.prevout.hash);
+                if (!from) {
+                    continue;
+                }
+                // The other source of inputs that is allowed is the coinbase
+                // transaction whose outputs matured in this block:
+                if (from->IsCoinBase() && (from->nHeight == (pindex->nHeight - COINBASE_MATURITY))) {
```

Comment:
This seems like a weird extra rule, but I actually see it as reducing inconsistencies from a user perspective. With forward block's extension outputs there are outputs yet-to-be-defined which need to be captured by the block-final transaction since they carry value but don't actually enter the UTXO set.

Consider, for example, a zcash-like anonymous/shielded spend: you send 5 coins to an output that you never intend to spend directly, but later you "claim" 5 coins by showing that you have the secret from the prior output, but without revealing which one, just that it's not among the ones that have not been revealed already. Now to keep those 5 coins in the pot of funds for you to claim later means the output needs to be spendable and collected in the block it appears by the block-final transaction, where the value is kept and carried over from block to block until you reclaim it.

Now what if, as a miner, you want to mine directly into a shielded output? The block-final transaction of the block you mine can't claim the funds because of coinbase maturity rules: it is the miner 100 blocks later that needs to. So on the back end we need to deal with some complexity: a block-final transaction can claim funds from either the same block or the coinbase 100 blocks prior. But from the user perspective this is consistent: you can create an extension output anywhere, in a user transaction or the coinbase, and in the block it matures it is claimed by the block-final transaction and acted upon.

Reply:
It occurs to me though that this line isn't compatible with the protocol cleanup hard fork, which reduces the coinbase maturity to 1 block. It's probably enough to check `<=` here, which I will modify it to do.

Reply:
Er, should be `>=`. I made the change and added a test.

Reply:
I understand now, thanks.